Remove overload of GPU max_pool3d with kernel_width; fix nan, inf in GPU {fractional,adaptive} max_pool{2,3}d

xwang233 opened the pull request:
Fix #39846.
Fix #39044

The problem was that `max_pool3d_with_indices_single_out_frame` has an overload of kernel_width being a template argument. The two overloaded kernels were supposed to be identical, however, they were not. 

The general version
https://github.com/pytorch/pytorch/blob/da3073e9b1db503f106842339f50f522d973be84/aten/src/ATen/native/cuda/DilatedMaxPool3d.cu#L69-L73

The overloaded version
https://github.com/pytorch/pytorch/blob/da3073e9b1db503f106842339f50f522d973be84/aten/src/ATen/native/cuda/DilatedMaxPool3d.cu#L130-L134

While the max_pool3d being "switch-case"-ed to the overloaded version, the NaN value comparison is ignored. Also, maintaining two overloaded versions of such a complicated kernel would be hard. I'm not sure if the overloaded version would even give huge performance benefit. So I propose to remove the kernel_width overloaded version.

Also, the current test of max_pool_XD_nan forgot the device kwarg. I added that. 

Edit: profiling before and after
script: https://github.com/xwang233/code-snippet/blob/master/maxpool-3d-kw-template-arg/a.py
plot: https://github.com/xwang233/code-snippet/blob/master/maxpool-3d-kw-template-arg/b.ipynb

The performance difference is within +- 5%.

xwang233 committed:
remove overload of max_pool3d with kernel_width

xwang233 committed:
test

xwang233 commented:
cc @ptrblck

xwang233 committed:
dtype, too

xwang233 committed:
Merge remote-tracking branch 'upstream/viable/strict' into max_pool_3d-fix-39846

ezyang commented:
@xwang233 it doesn't have to be comprehensive, but could you do a quick before/after performance comparison? Thanks!

ngimel commented:
Could you also roll the fix to #39044 here? I suspect it's a very similar issue.

ngimel commented:
Agree with @ezyang, usually template arguments provide a significant perf benefits to kernels like this, so it might make sense to put in a one-line fix to the original kernel, depending on the results of your perf tests.

xwang233 committed:
propagate nan

xwang233 committed:
test fractional maxpool_XD

xwang233 commented:
Thanks for the comments. I did some profilings for max_pool3d before and after this PR. The script and plot are available in the links in the updated original post. The performance difference is less than +- 5%, within fluctuation range. 

From my experience 
```c++
int wEnd = min(wStart + (KERNEL_WIDTH - 1) * dilationW + 1, iwidth);
...
for (int w = wStart; w < wEnd; w += dilationW)
```
`kernel_width` is multiplied by a runtime variable (`dilationW`) and wrapped by a `min` function, which makes compiler optimization much less effective. A much better optimization is that we can write a specialized kernel with all default values `dilation=1`, `padding=0`, `stride=kernel_size`, and that can give a lot performance gain. However, this also increases the maintenance cost.

xwang233 commented:
I also updated the nan value handling in fractional max pool 2d and 3d. Tests are also added. However, CPU fractional maxpool yields an internal error from this test
```
RuntimeError: maxVal != -std::numeric_limits<scalar_t>::infinity() INTERNAL ASSERT FAILED at 
"../aten/src/ATen/native/FractionalMaxPool2d.cpp":83, please report a bug to PyTorch. 
```
So I make the test `onlyCUDA`. I can open another issue for this CPU failure.

xwang233 committed:
Merge remote-tracking branch 'upstream/viable/strict' into max_pool_3d-fix-39846

ngimel commented:
Cool, can you also please handle fractional_max_pool issue with -inf-filled tensors, and remove asserts from the kernel (because they are ignored anyway)?

xwang233 committed:
remove assert

ngimel commented:
maxval is still set as numeric_limits::lowest, which is wrong if your pooling window has only -inf and/or numeric_limits::lowest.

xwang233 commented:
Thanks! I'm testing on it.

xwang233 committed:
float -inf

xwang233 committed:
test pooling -inf

xwang233 commented:
I fixed those `-inf` from `{adaptive, fractional} max pooling` and added tests. The CPU `adaptive_max_pool` does not propagate `-inf`, so I left a TODO there.

xwang233 committed:
lint

ngimel commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -77,7 +77,7 @@ __global__ void adaptivemaxpool(T *input, T *output, int64_t *indices,
       int64_t *ptr_ind = indices + oh*osizeW + ow;
       int argmax = -1;
-      T max = THCNumerics<T>::min();
+      T max = at::numeric_limits<T>::lower_bound(); // -Infinity
```
If there are only -inf in the pooling window, argmax will not be assigned, and there will be invalid memory access in backward. Your tests don't test backward.

xwang233 replied:
Fixed. `argmax` is now initialized with a value similar to those in the loop.

ngimel reviewed (changes requested):
Can you also please file an issue for CPU errors?

xwang233 committed:
default index

xwang233 committed:
test

xwang233 committed:
Merge remote-tracking branch 'upstream/viable/strict' into max_pool_3d-fix-39846

xwang233 commented:
ROCm failure seems real. Do you mind if I add `@skipCUDAIfRocm`?
```
01:31:26 ======================================================================
01:31:26 ERROR: test_fractional_max_pool_nan_inf_cuda_bfloat16 (__main__.TestNNDeviceTypeCUDA)
01:31:26 ----------------------------------------------------------------------
01:31:26 Traceback (most recent call last):
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 769, in wrapper
01:31:26     method(*args, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 241, in instantiated_test
01:31:26     result = test(self, device_arg, dtype)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 506, in only_fn
01:31:26     return fn(slf, device, *args, **kwargs)
01:31:26   File "test_nn.py", line 10979, in test_fractional_max_pool_nan_inf
01:31:26     res = fn(x)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 609, in _call_impl
01:31:26     result = self.forward(*input, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/modules/pooling.py", line 750, in forward
01:31:26     _random_samples=self._random_samples)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/_jit_internal.py", line 243, in fn
01:31:26     return if_false(*args, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/functional.py", line 410, in _fractional_max_pool2d
01:31:26     _random_samples)[0]
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/functional.py", line 395, in fractional_max_pool2d_with_indices
01:31:26     return torch._C._nn.fractional_max_pool2d(input, kernel_size, output_size, _random_samples)
01:31:26 RuntimeError: "fractional_max_pool2d_out_cuda_frame" not implemented for 'BFloat16'
```

ngimel commented:
Yes, please change dtypesIfCUDA

xwang233 committed:
remove bf16 in dtypes

xwang233 committed:
Merge remote-tracking branch 'upstream/viable/strict' into max_pool_3d-fix-39846
